p4-16/spec/P4-16-spec.mdk

```diff
@@ -6416,6 +6416,160 @@ ck16.update( { h.ipv4.ttl, h.ipv4.proto } );
+| control     |  no       |  no     | no      | no      |
+| extern      | yes       | yes     | yes     | no      |
+| table       |  no       |  no     | no      | no      |
+| value types | yes       | yes     | yes     | yes     |
+|-------------|-----------|---------|---------|---------|
+
+The restrictions on return types of extern methods is the same as for
+parameters, with the addition that `void` can be a return type for
+methods that return no value.
+
+The next table lists restrictions on what kinds of calls can be made
+from which places in a P4 program.  Calling a parser, control, or
+table means invoking its `apply()` method.  The row for `extern`
+describes where extern method calls can be made from.
+
+|-------------|-------------------------------------------------|
+|             | can be called at run time from this place in a P4 program |||||
+|             |-----------|---------|---------|------------|--------|
+|             |           | control | action  | action not |        |
+|             | parser    | apply   | used by | used by    |        |
+| This type   | state     | block   | table   | any table  | extern |
++-------------+:---------:+:-------:+:-------:+:----------:+:------:+
+| package     | N/A       | N/A     | N/A     | N/A        | N/A    |
+| parser      | yes       |  no     | no      | no         | no     |
+| control     |  no       | yes     | no      | yes        | no     |
+| extern      | yes       | yes     | yes     | yes        | no     |
+| table       |  no       | yes     | no      | yes        | no     |
+| action      |  no       | yes     | yes     | yes        | no     |
+| value types | N/A       | N/A     | N/A     | N/A        | N/A    |
+|-------------|-----------|---------|---------|------------|--------|
+
+There may not be any recursion in calls, neither by a thing calling
+itself directly, nor mutual recursion.
+
+An extern can never cause any other type of P4 program object to be
+called.  See Section [#sec-calling-convention-justification].
+
+An "action used by a table" means an action that is either in the
+`actions` list of a table, or it is possible for the action to be
+called from such an action, perhaps through one or more intermediate
+action calls.
+
+Any action can be called from a control `apply` block.
+
+It is allowed to write actions that are not used by any table.  Such
+"actions not used by any table" have exactly the same restrictions on
+what they can call as a control `apply` block does.  A potential
+convenience of creating such an action is that it can be defined
+within the control that calls it, giving the action access to all
+parameter values and local variables defined earlier within its
+enclosing control.  By contrast, control definitions cannot be nested.
+
+Note that while the extern row shows that extern methods can be called
+from many places, particular externs may have additional restrictions
```

Comment:
I've added a few words regarding this that may address your comment.  Take a look at the updated version to see.